python/ray/util/sgd/v2/trainer.py

```diff
@@ -132,41 +153,147 @@ def run(self,
             list corresponds to the output of the training function from
             each worker.
         """
-        train_func = self._get_train_func(train_func, config)
         # TODO(matt): Set default callbacks.
         callbacks = [] if callbacks is None else callbacks
         finished_with_errors = False
 
-        try:
-            for callback in callbacks:
-                callback.start_training()
-            self._executor.start_training(train_func, checkpoint)
+        for callback in callbacks:
+            callback.start_training()
 
-            while True:
-                intermediate_results = self._executor.fetch_next_result()
-                if intermediate_results is None:
-                    break
-                else:
-                    for callback in callbacks:
-                        callback.handle_result(intermediate_results)
-
-            return self._executor.finish_training()
-        except InactiveWorkerGroupError:
-            finished_with_errors = True
-            raise RuntimeError(
-                "This Trainer is not active. It is either shutdown already or "
-                "never started in the first place. Either create a new "
-                "Trainer or start this one.") from None
-        except SGDBackendError:
-            finished_with_errors = True
-            raise RuntimeError("Training failed. You should not be seeing "
-                               "this error and this is a bug. Please create "
-                               "a new issue at "
-                               "https://github.com/ray-project/ray.") from None
+        try:
+            iterator = self.run_iterator(
+                train_func=train_func, config=config, checkpoint=checkpoint)
+            for intermediate_result in iterator:
+                for callback in callbacks:
+                    callback.handle_result(intermediate_result)
+
+            assert iterator.is_finished()
+            return iterator.get_returns()
         finally:
             for callback in callbacks:
                 callback.finish_training(error=finished_with_errors)
 
+    def run_iterator(
+            self,
+            train_func: Union[Callable[[], T], Callable[[Dict[str, Any]], T]],
+            config: Optional[Dict[str, Any]] = None,
+            checkpoint: Optional[Dict] = None) -> Iterator[List[Any]]:
```

Comment:
This will always iterate over `Dict`s right?
```suggestion
            checkpoint: Optional[Dict] = None) -> Iterator[List[Dict]]:
```

Reply:
Updated!